packages/server/lib/util/socket_whitelist.ts

```diff
@@ -0,0 +1,46 @@
+import _ from 'lodash'
+import Debug from 'debug'
+import net from 'net'
+import { Request } from 'express'
+
+const debug = Debug('cypress:server:util:socket_whitelist')
+
+/**
+ * Utility to validate incoming, local socket connections against a list of
+ * expected client TCP ports.
+ */
+export class SocketWhitelist {
+  whitelistedLocalPorts: number[] = []
+
+  /**
+   * Add a socket to the whitelist.
+   */
+  add = (socket: net.Socket) => {
+    const { localPort } = socket
+
+    debug('whitelisting socket %o', { localPort })
+    this.whitelistedLocalPorts.push(localPort)
+
+    socket.once('close', () => {
+      debug('whitelisted socket closed, removing %o', { localPort })
+      this._remove(socket)
+    })
+  }
+
+  _remove (socket: net.Socket) {
+    _.remove(this.whitelistedLocalPorts, socket.localPort)
+  }
+
+  /**
+   * Is this socket that this request originated from whitelisted?
+   */
+  isRequestWhitelisted (req: Request) {
+    const { remotePort, remoteAddress } = req.socket
+    const isWhitelisted = this.whitelistedLocalPorts.includes(remotePort!)
+      && ['127.0.0.1', '::1'].includes(remoteAddress!)
```

Comment:
are the localhost ips necessary here? i thought we only bound to `loopback` - or is it here for extra protection

Reply:
yeah, this is just explicitly asserting what is implicitly true already